storage/local/storage.go

```diff
@@ -360,11 +371,28 @@ func (s *memorySeriesStorage) GetMetricForFingerprint(fp clientmodel.Fingerprint
 
 // AppendSamples implements Storage.
 func (s *memorySeriesStorage) AppendSamples(samples clientmodel.Samples) {
+	if len(samples) == 0 {
+		return
+	}
+
+	s.appendMtx.Lock()
+	defer s.appendMtx.Unlock()
+
 	for _, sample := range samples {
-		s.appendSample(sample)
+		if sample.Timestamp != s.lastTimestampAppended {
+			// Timestamp has changed. We have to wait for all
+			// appendSample to complete before proceeding.
+			s.appendWaitGroup.Wait()
+			s.lastTimestampAppended = sample.Timestamp
+		}
+
+		s.appendWaitGroup.Add(1)
+		go func(sample *clientmodel.Sample) {
```

Comment:
I thought about this a lot and played around with it... With a worker pool, you want them to road from a channel, but our interface design read's from the channel in main.go (to then send the samples to both, the local and the remote storage). Here we needed to implement a dispatcher or something, hand stuff over to the pool. That's probably slower than creating a goroutine on demand. It's pretty common practice to create very ephemeral goroutines, even in the stdlib. Goroutine creation is cheap for a reason.

The code here is proven to be way faster than the previous version (under heavy load). It was like day and night on my machine.